pkg/pillar/hypervisor/kvm.go

```diff
@@ -457,23 +457,138 @@ func (ctx kvmContext) Name() string {
-	} else {
-		return ctx
 	}
+	return ctx
 }
 
-// TBD: Have a better way to calculate this number.
-// For now it is based on some trial-and-error experiments.
-// Container limit is reduced to 100MiB.
-func minVMMOverhead(config types.DomainConfig) int64 {
+func estimatedVMMOverhead(domainName string, config types.DomainConfig, aa *types.AssignableAdapters) (int64, error) {
+	// Container limit is set to 100MiB.
+	// This is a temporary solution until we have a better way to predict
+	// the memory usage of the container.
 	if config.IsOCIContainer() {
-		return 100 << 20 // Mb in bytes
+		return 100 << 20, nil // Mb in bytes
+	}
+
+	var overhead int64
+
+	mmioOverhead, err := mmioVMMOverhead(domainName, config, aa)
+
+	if err != nil {
+		return 0, logError("mmioVMMOverhead() failed for domain %s: %v",
+			domainName, err)
 	}
-	return 600 << 20 // Mb in bytes
+	overhead = undefinedVMMOverhead() + ramVMMOverhead(config) +
+		qemuVMMOverhead() + cpuVMMOverhead(config) + mmioOverhead
+
+	return overhead, nil
+}
+
+func ramVMMOverhead(config types.DomainConfig) int64 {
+	// 2.5% of the total RAM allocated for VM in bytes
+	// this formula is precise and well explained in the following article:
+	// TBD
+	return int64(config.Memory) * 1024 * 25 / 1000
+}
+
+// overhead for qemu binaries and libraries
+func qemuVMMOverhead() int64 {
+	return 20 << 20 // Mb in bytes
 }
 
-func vmmOverhead(config types.DomainConfig,
-	globalConfig *types.ConfigItemValueMap) (int64, error) {
+// overhead for VMM memory mapped IO
+// it fluctuates between 0.66 and 0.81 % of MMIO total size
+// for all mapped devices. Set it to 1% to be on the safe side
+// this can be a pretty big number for GPUs with very big
+// aperture size (e.g. 64G for NVIDIA A40)
+func mmioVMMOverhead(domainName string, config types.DomainConfig, aa *types.AssignableAdapters) (int64, error) {
+	var pciAssignments []pciDevice
+	var mmioSize uint64
+
+	for _, adapter := range config.IoAdapterList {
+		logrus.Debugf("processing adapter %d %s\n", adapter.Type, adapter.Name)
+		aaList := aa.LookupIoBundleAny(adapter.Name)
+		// We reserved it in handleCreate so nobody could have stolen it
+		if len(aaList) == 0 {
+			return 0, logError("IoBundle disappeared %d %s for %s\n",
+				adapter.Type, adapter.Name, domainName)
+		}
+		for _, ib := range aaList {
+			if ib == nil {
+				continue
+			}
+			if ib.UsedByUUID != config.UUIDandVersion.UUID {
+				return 0, logError("IoBundle not ours %s: %d %s for %s\n",
+					ib.UsedByUUID, adapter.Type, adapter.Name,
+					domainName)
+			}
+			if ib.PciLong != "" && ib.UsbAddr == "" {
+				logrus.Infof("Adding PCI device <%s>\n", ib.PciLong)
+				tap := pciDevice{pciLong: ib.PciLong, ioType: ib.Type}
+				pciAssignments = addNoDuplicatePCI(pciAssignments, tap)
+			}
+		}
+	}
+
+	for _, dev := range pciAssignments {
+		logrus.Infof("PCI device %s %d\n", dev.pciLong, dev.ioType)
+		// read the size of the PCI device aperture
+		if dev.ioType != types.IoOther {
```

Comment:
Why don't you want to look at devices of type vga and hdmi?significant today (but don't know about smart nics down the road). But only looking at IoOther seems odd.

Reply:
I agree, potentially we should handle all PCI devices here. I was under false impression that all GPUs are fall into IoOther type. 
We should also improve a way how we detect device class in spec.sh. Now we just `lspci -D -s "${pci}" | grep -q VGA` and this is not correct. We should look at 
```
cat /sys/bus/pci/devices/0000:01:00.0/class
0x030200
```
and decode a device class according to PCI specification

![image](https://github.com/lf-edge/eve/assets/88966866/fdbb12cc-5f58-45ae-8d74-3696bea1f741)

Reply:
I agree we can and should improve spec.sh (and we might also want to define more IoTypes as part of that based on the classes e.g., for 12h accelerators etc.).

But for my understanding, do you see any downside to doing these checks/calculations for all classes?